src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,205 @@
+      });
+    co_return enrich(v);
+}
+
+ss::future<self_test_frontend::node_test_state>
+self_test_frontend::remote_invoke::stop_test() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        /// Bigger timeout to wait for executing jobs to shutdown
+        return c.stop_test(
+          empty_request{}, rpc::client_opts(raft::clock_type::now() + 30s));
+    });
+    co_return enrich(v);
+}
+
+ss::future<self_test_frontend::node_test_state>
+self_test_frontend::remote_invoke::get_status() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        return c.get_status(
+          empty_request{}, rpc::client_opts(raft::clock_type::now() + 5s));
+    });
+    co_return enrich(v);
+}
+
+/// Implementation of methods in self_test_frontend
+
+self_test_frontend::self_test_frontend(
+  model::node_id self,
+  ss::sharded<cluster::members_table>& members,
+  ss::sharded<cluster::self_test_backend>& self_be,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _members(members)
+  , _self_be(self_be)
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<uuid_t> self_test_frontend::start_test(
+  std::optional<diskcheck_opts> dto, std::optional<netcheck_opts> nto) {
+    auto gate_holder = _gate.hold();
+    if (!dto && !nto) {
+        throw self_test_exception("No tests specified to run");
+    }
+    const auto stopped_results = co_await stop_test();
+    if (!stopped_results.finished()) {
+        vlog(
+          clusterlog.warn,
+          "Could not stop already running test before calling start,  "
+          "sending start RPCs out anyway.");
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    auto state = co_await invoke_on_all_nodes(
+      [req](auto& handle) { return handle->start_test(req); });
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
+    while (!test_state.finished() && retries-- > 0) {
+        /// invoke_on_all will filter brokers that have shut down cleanly
+        /// from its participants list, retries (if occur) will only occur
+        /// on previously unreachable nodes
+        test_state = co_await invoke_on_all_nodes(
+          [](auto& handle) { return handle->stop_test(); });
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state.active_participant_ids(),
+              retries);
```

Comment:
nit: this is a little confusing since we're not actually showing all the participants. How about something like

"Failed to stop self tests; remaining running tests at nodes {active_participant_ids}, remaining retries {retries}"

Reply:
I decided to print only the participants for which contain the stuck jobs

Reply:
Right, I just think it's worth updating the wording to make it clear that the printed participants are the _remaining_ ones, not _all_ participants